From 07f2f06e95cba18cf1308c43363633fb120f1e0d Mon Sep 17 00:00:00 2001
From: Bogdan Hamciuc <bogdan.hamciuc@freescale.com>
Date: Thu, 30 May 2013 18:05:56 +0300
Subject: [PATCH 214/987] dpaa_eth: Create TxConf FQs with HOLDACTIVE flag

[Original patch taken from QorIQ-SDK-V1.6-SOURCE-20140619-yocto.iso]

Tx Confirmation FQs tend, with originating traffic, to generate a high
number of portal interrupts. A machine doing only TCP transmits may well
end up having more portal interrupts than its Rx peer.

By setting the HOLDACTIVE flag on the Tx Confirmation FQs, we
significantly reduce the number of portal interrupts in termination
tests.
The quantitative effect is anywhere between 2% and 4% on Tx-mainly
cores (most visible in UDP_STREAM tests, least visible on TCP_STREAM).

Signed-off-by: Bogdan Hamciuc <bogdan.hamciuc@freescale.com>
Change-Id: I54e2e8287f72da0cfe231af91c624be805ddee17
Reviewed-on: http://git.am.freescale.net:8181/2815
Tested-by: Review Code-CDREVIEW <CDREVIEW@freescale.com>
Reviewed-by: Sovaiala Cristian-Constantin-B39531 <Cristian.Sovaiala@freescale.com>
Reviewed-by: Bucur Madalin-Cristian-B32716 <madalin.bucur@freescale.com>
Reviewed-by: Radulescu Ruxandra Ioana-B05472 <ruxandra.radulescu@freescale.com>
Reviewed-by: Fleming Andrew-AFLEMING <AFLEMING@freescale.com>
Tested-by: Fleming Andrew-AFLEMING <AFLEMING@freescale.com>
---
 drivers/net/ethernet/freescale/dpa/dpaa_eth.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/net/ethernet/freescale/dpa/dpaa_eth.c b/drivers/net/ethernet/freescale/dpa/dpaa_eth.c
index 2f6b80f..75cd26f 100644
--- a/drivers/net/ethernet/freescale/dpa/dpaa_eth.c
+++ b/drivers/net/ethernet/freescale/dpa/dpaa_eth.c
@@ -611,6 +611,14 @@ _dpa_fq_alloc(struct list_head *list, struct dpa_fq *dpa_fq)
 		/* FIXME: why would we want to keep an empty FQ in cache? */
 		initfq.fqd.fq_ctrl = QM_FQCTRL_PREFERINCACHE;
 
+#ifdef CONFIG_FSL_DPAA_ETH_SG_SUPPORT
+		/* Try to reduce the number of portal interrupts for
+		 * Tx Confirmation FQs.
+		 */
+		if (dpa_fq->fq_type == FQ_TYPE_TX_CONFIRM)
+			initfq.fqd.fq_ctrl |= QM_FQCTRL_HOLDACTIVE;
+#endif
+
 		/* FQ placement */
 		initfq.we_mask |= QM_INITFQ_WE_DESTWQ;
 
-- 
1.9.1

